File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/state/MatchesViewState.java
package com.benoitquenaudon.tvfoot.red.app.domain.matches.state;

import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.HeaderRowDisplayable;
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.LoadingRowDisplayable;
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.MatchRowDisplayable;
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.MatchesItemDisplayable;
import com.google.auto.value.AutoValue;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import javax.annotation.Nullable;

@AutoValue public abstract class MatchesViewState {
  public List<MatchesItemDisplayable> matchesItemDisplayables() {
    List<String> headers = new ArrayList<>();
    List<MatchesItemDisplayable> items = new ArrayList<>();
    for (MatchRowDisplayable match : matches()) {
      if (!headers.contains(match.headerKey())) {
        headers.add(match.headerKey());
        items.add(HeaderRowDisplayable.create(match.headerKey()));
      }
      items.add(match);
    }
    if (!items.isEmpty()) {
      items.add(LoadingRowDisplayable.create());
    }
    return items;
  }

  public abstract List<MatchRowDisplayable> matches();

  @Nullable public abstract Throwable error();

  public abstract boolean nextPageLoading();

  public abstract boolean refreshLoading();

  public abstract Integer currentPage();

  public static Builder builder() {
    return new AutoValue_MatchesViewState.Builder().matches(Collections.emptyList())
        .error(null)
        .nextPageLoading(false)
        .refreshLoading(false)
        .currentPage(0);
  }

  public abstract Builder buildWith();

  static MatchesViewState idle() {
    return MatchesViewState.builder().build();
  }

  @AutoValue.Builder public static abstract class Builder {
    public abstract Builder matches(List<MatchRowDisplayable> matches);

    public abstract Builder error(@Nullable Throwable error);

    public abstract Builder nextPageLoading(boolean nextPageLoading);

    public abstract Builder refreshLoading(boolean pullToRefreshLoading);

    public abstract Builder currentPage(Integer CurrentPage);

    public abstract MatchesViewState build();
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/MatchesViewState.kt
package com.benoitquenaudon.tvfoot.red.app.domain.matches

import com.benoitquenaudon.tvfoot.red.app.data.entity.FilterTeam
import com.benoitquenaudon.tvfoot.red.app.data.entity.Tag
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.HeaderRowDisplayable
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.LoadingRowDisplayable
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.MatchRowDisplayable
import com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable.MatchesItemDisplayable
import com.benoitquenaudon.tvfoot.red.app.domain.matches.filters.FiltersItemDisplayable.TeamSearchResultDisplayable
import com.benoitquenaudon.tvfoot.red.app.mvi.MviViewState
import com.benoitquenaudon.tvfoot.red.util.TagName
import com.benoitquenaudon.tvfoot.red.util.TagTargets
import com.benoitquenaudon.tvfoot.red.util.TeamCode
import java.util.ArrayList

data class MatchesViewState(
    val matches: List<MatchRowDisplayable> = emptyList(),
    val error: Throwable? = null,
    val nextPageLoading: Boolean = false,
    val refreshLoading: Boolean = false,
    val teamMatchesLoading: Boolean = false,
    val currentPage: Int = 0,
    val hasMore: Boolean = true,
    val tagsLoading: Boolean = false,
    val tagsError: Throwable? = null,
    val tags: List<Tag> = emptyList(),
    var filteredCompetitions: Map<TagName, TagTargets> = emptyMap(),
    var filteredBroadcasters: Map<TagName, TagTargets> = emptyMap(),
    val searchInput: String = "",
    val searchingTeam: Boolean = false,
    val searchedTeams: List<TeamSearchResultDisplayable> = emptyList(),
    val teams: List<FilterTeam> = emptyList(),
    val filteredTeams: List<TeamCode> = emptyList()
) : MviViewState {
  fun matchesItemDisplayables(
      nextPageLoading: Boolean,
      loadingSpecificMatches: Boolean,
      filteredBroadcasters: Map<TagName, TagTargets>,
      filteredCompetitions: Map<TagName, TagTargets>,
      filteredTeams: List<TeamCode>
  ): List<MatchesItemDisplayable> {

    val headers = ArrayList<String>()
    val items = ArrayList<MatchesItemDisplayable>()

    val filteredCompetitionTargets = filteredCompetitions.values.flatten().toSet()
    val filteredBroadcasterTargets = filteredBroadcasters.values.flatten().toSet()
    val filteredTeamsCodes = filteredTeams.toSet()
    for (match in matches) {
      if (filteredCompetitionTargets.isNotEmpty() && filteredCompetitionTargets.intersect(
          match.tags).isEmpty()) {
        continue
      }
      if (filteredBroadcasterTargets.isNotEmpty() &&
          filteredBroadcasterTargets.intersect(
              match.broadcasters.map { it.broadcasterCode }).isEmpty()) {
        continue
      }
      if (filteredTeamsCodes.isNotEmpty() &&
          filteredTeamsCodes.intersect(
              setOf(match.homeTeam.code, match.awayTeam.code)).isEmpty()) {
        continue
      }

      if (!headers.contains(match.headerKey)) {
        headers.add(match.headerKey)
        items.add(HeaderRowDisplayable.create(match.headerKey))
      }
      items.add(match)
    }
    if (!items.isEmpty() && (nextPageLoading || loadingSpecificMatches)) {
      items.add(LoadingRowDisplayable)
    }
    return items
  }

  override fun toString(): String {
    return "MatchesViewState(nextPageLoading=$nextPageLoading, refreshLoading=$refreshLoading, teamMatchesLoading=$teamMatchesLoading, currentPage=$currentPage, hasMore=$hasMore, tagsLoading=$tagsLoading, searchInput='$searchInput', searchingTeam=$searchingTeam)"
  }


  companion object {
    fun idle(): MatchesViewState = MatchesViewState()
  }
}
